### DISCUSSION TITLE
How to get Global keyboard and mouse listeners

### DISCUSSION THREAD

**Discussion by Aswathips** [AvaloniaUI/Avalonia#8071] [Q&A]
Hello,
Is there is any library that provide global keyboard and mouse listeners for dot net, just like the jnativehook library in java.
or how can we detect the system idle time . i want the code to run on Windows , Linux and Mac . 

Also anyone please give me any suggestions to load the avalonia application to the systemtray.
Please help me.

**Comment by maxkatz6** [comment]
You can get some global input from InputManager class. But if you need to get input from the OS globally, you need to do it by yourself. Maybe there are some alternatives or ports of  jnativehook.

**Comment by sandreas** [comment]
Maybe SharpHook would help you...
https://github.com/TolikPylypchuk/SharpHook

**Comment by xmaxrayx** [comment]
had wasted my time for 3+ hours trying figurant sadly this "cross-platform" issue you get Linux devs that want their OS nightmare to port while in windows you can just use user32.dll for ages and never break.

current time I just wrote for windows with ` #define Win `so  maybe after 5-10years linux devs will agree to use 1-standerd and after that i will add `#define linux` and custom hock for them not worh it now.

**Reply by thevortexcloud** [reply]
Linux really only has two major desktop standards? X11, which works pretty closely to Windows and Wayland which works nothing like Windows. X11 is rapidly being deprecated in favour of Wayland, and Wayland itself supports X11 via XWayland. The current recommended way to do global shortcuts on Linux is to use a portal, which should work on most desktops (and Wayland) without having to reinvent the wheel for every distro.

https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.GlobalShortcuts.html

If you don't have much experience working with Linux (which it sounds like you don't) finding this stuff can be hard. But the information is there.

Wayland and most of the Linux world is trying to move in direction of better security. Part of that means avoiding giving direct keyboard access when it's not needed (as there is not much difference between a key logger and global keyboard shortcut). Trying to do things exactly the same as on Windows is going to make things painful for yourself.
